## Загрузка и подготовка данных

In [78]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
import numpy as np
from scipy import stats as st

Загрузка файлов

In [79]:
geo0 = pd.read_csv('/datasets/geo_data_0.csv')
geo1 = pd.read_csv('/datasets/geo_data_1.csv')
geo2 = pd.read_csv('/datasets/geo_data_2.csv')

In [80]:
display(geo0)
print(geo0.info())

id        f0        f1        f2     product
0      txEyH  0.705745 -0.497823  1.221170  105.280062
1      2acmU  1.334711 -0.340164  4.365080   73.037750
2      409Wp  1.022732  0.151990  1.419926   85.265647
3      iJLyR -0.032172  0.139033  2.978566  168.620776
4      Xdl7t  1.988431  0.155413  4.751769  154.036647
...      ...       ...       ...       ...         ...
99995  DLsed  0.971957  0.370953  6.075346  110.744026
99996  QKivN  1.392429 -0.382606  1.273912  122.346843
99997  3rnvd  1.029585  0.018787 -1.348308   64.375443
99998  7kl59  0.998163 -0.528582  1.583869   74.040764
99999  1CWhH  1.764754 -0.266417  5.722849  149.633246

[100000 rows x 5 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None


In [81]:
print(geo2['f0'].max())
print(geo2['f0'].min())

7.23826247979405
-8.760003624213763


In [82]:
display(geo1)
print(geo1.info())

id         f0         f1        f2     product
0      kBEdx -15.001348  -8.276000 -0.005876    3.179103
1      62mP7  14.272088  -3.475083  0.999183   26.953261
2      vyE1P   6.263187  -5.948386  5.001160  134.766305
3      KcrkZ -13.081196 -11.506057  4.999415  137.945408
4      AHL4O  12.702195  -8.147433  5.004363  134.766305
...      ...        ...        ...       ...         ...
99995  QywKC   9.535637  -6.878139  1.998296   53.906522
99996  ptvty -10.160631 -12.558096  5.005581  137.945408
99997  09gWa  -7.378891  -3.084104  4.998651  137.945408
99998  rqwUm   0.665714  -6.152593  1.000146   30.132364
99999  relB0  -3.426139  -7.794274 -0.003299    3.179103

[100000 rows x 5 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None


In [83]:
print(geo2['f1'].max())
print(geo2['f1'].min())

7.844801270084258
-7.084019760867246


In [84]:
display(geo2)
print(geo2.info())

id        f0        f1        f2     product
0      fwXo0 -1.146987  0.963328 -0.828965   27.758673
1      WJtFt  0.262778  0.269839 -2.530187   56.069697
2      ovLUW  0.194587  0.289035 -5.586433   62.871910
3      q6cA6  2.236060 -0.553760  0.930038  114.572842
4      WPMUX -0.515993  1.716266  5.899011  149.600746
...      ...       ...       ...       ...         ...
99995  4GxBu -1.777037  1.125220  6.263374  172.327046
99996  YKFjq -1.261523 -0.894828  2.524545  138.748846
99997  tKPY3 -1.199934 -2.957637  5.219411  157.080080
99998  nmxp2 -2.419896  2.417221 -5.548444   51.795253
99999  V9kWn -2.551421 -2.025625  6.090891  102.775767

[100000 rows x 5 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None


In [85]:
print(geo2['f2'].max())
print(geo2['f2'].min())

16.73940205869952
-11.97033454147699


### Вывод

Данные загружены. Пропусков не обнаружено.

## Обучение и проверка модели

In [86]:
model = LinearRegression()

Созданим функцию для обучения модели

In [87]:
def predict_target (data):
    features = data.drop(['id', 'product'], axis = 1)
    target = data['product']
    features_train, features_valid, target_train, target_valid = train_test_split(features, target, test_size=.25, random_state=666)
    model.fit(features_train, target_train)
    return pd.Series(model.predict(features_valid)), target_valid

In [88]:
predicted_0, target_0 = predict_target(geo0)

In [89]:
predicted_1, target_1 = predict_target(geo1)

In [90]:
predicted_2, target_2 = predict_target(geo2)

Создадим функцию для рассчета RMSE

In [91]:
def rmse (answers, predictions):
    mse = mean_squared_error(answers, predictions)
    print('Средний запас сырья:', predictions.mean())
    print('RMSE:', mse ** 0.5)

In [92]:
rmse(target_0, predicted_0)

Средний запас сырья: 92.59883747361135
RMSE: 37.78852914975086


In [93]:
rmse(target_1, predicted_1)

Средний запас сырья: 69.11960454214399
RMSE: 0.8899276270922486


In [94]:
rmse(target_2, predicted_2)

Средний запас сырья: 95.05831290258574
RMSE: 40.008583723061925


## Подготовка к расчёту прибыли

Посчитаем минимальное содержание для безубыточности 

In [95]:
BUDGET = 10**10
COUNT = 200
COST = 450000

In [96]:
min_produce = BUDGET/(COUNT*COST)
print('минимальное среднее количество, чтобы проект был прибыльным: '+str(min_produce))

минимальное среднее количество, чтобы проект был прибыльным: 111.11111111111111


In [97]:
def profit(target, probabilities):
    target = pd.Series(target).reset_index(drop=True)
    probabilities = pd.Series(probabilities).reset_index(drop=True)
    probs_sorted = probabilities.sort_values(ascending=False)
    selected = target[probs_sorted.index][:200]
    return ((COST * selected.sum()) - BUDGET) / 1000000000

## Расчёт прибыли и рисков 

In [98]:
state = np.random.RandomState(12345)

In [99]:
def bootstrap (target, predictions):
    values = []
    for i in range(1000):
        target_subsample = target.reset_index(drop=True).sample(500, replace=True, random_state=state)
        probs_subsample = predictions[target_subsample.index]
        rev = profit(target_subsample, probs_subsample)
        values.append(rev)

    values = pd.Series(values)
    lower = values.quantile(0.05)
    confidence_interval = (values.quantile(0.025), values.quantile(0.975))
    mean = values.mean()

    print("Средняя выручка:", mean)
    print("Доверительный интервал:", confidence_interval)
    print("Риски убытка:", (values < 0).mean())

In [100]:
bootstrap(target_0, predicted_0)

Средняя выручка: 0.43069839930828546
Доверительный интервал: (-0.10671743695636005, 0.9307130825630421)
Риски убытка: 0.052


In [101]:
bootstrap(target_1, predicted_1)

Средняя выручка: 0.48536834632742604
Доверительный интервал: (0.08284952787924067, 0.8787523728795791)
Риски убытка: 0.005


In [102]:
bootstrap(target_2, predicted_2)

Средняя выручка: 0.3484111749716312
Доверительный интервал: (-0.18284438789284505, 0.8567012626397866)
Риски убытка: 0.098


## Вывод общий

- Средний необходимый минимум запасов для точки безубыточности 111.11
- При применении техники bootstrap:
- Оценка средней прибыли максимальна для региона 1 (480.91 млн.).
- Риск убытков для региона 1 (1.3%)
- Рекомендуем регион 1 для разработки.